pkg/web_app/lib/src/widget/dismissible/widget.dart

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:js_interop';
+
+import 'package:collection/collection.dart';
+import 'package:web/web.dart';
+
+import '../../web_util.dart';
+
+/// Forget dismissed messages that are more than 2 years old
+late final _deadline = DateTime.now().subtract(Duration(days: 365 * 2));
+
+/// Don't save more than 50 entries
+const _maxMissedMessages = 50;
+
+/// Create a dismissible widget on [element].
+///
+/// A `data-dismissible-by` is required, this must be a CSS selected for a
+/// child element that when clicked will dismiss the message by removing
+/// [element].
+///
+/// A hash is computed from `innerText` of [element], once dismissed this hash
+/// will be stored in `localStorage`. And next time this widget is instantiated
+/// with the same `innerText` it'll be removed immediately.
+///
+/// Hashes of dismissed messages will be stored for up to 2 years.
+/// No more than 50 dismissed messages are retained in `localStorage`.
+void create(HTMLElement element, Map<String, String> options) {
+  final by = options['by'];
+  if (by == null) throw UnsupportedError('data-dismissible-by required');
+
+  late final hash = _cheapNaiveHash(element.innerText);
+  if (_dismissed.any((e) => e.hash == hash)) {
+    element.remove();
+    return;
+  } else {
+    element.style.display = 'revert';
```

Comment:
Nit: the current content security policies allow this, but we had a plan to lock down the page by not allowing any such style update. I'm not sure if it is possible with the completion widget's positioning, but in this case, we could have a class that makes this widget visible.

Reply:
Are there CSP policies to prevent messing with style on individual elements?

I made it into a `dismissed` class instead, that seems pretty reasonable.